opal/mca/accelerator/base/accelerator_base_select.c

```diff
@@ -0,0 +1,148 @@
+#include "opal/constants.h"
+#include "opal/mca/base/base.h"
+#include "opal/mca/mca.h"
+#include "opal/mca/accelerator/base/base.h"
+#include "opal/mca/accelerator/accelerator.h"
+
+typedef struct accelerator_opened_component_t {
+  opal_list_item_t super;
+  opal_accelerator_base_component_t *accelerator_component;
+} accelerator_opened_component_t;
+
+
+int opal_accelerator_base_select(void)
+{
+
+    int priority = 0, best_priority = -1, ret = 0;
+    opal_list_item_t *item = NULL;
+    mca_base_component_list_item_t *cli = NULL;
+    opal_accelerator_base_component_t *component = NULL, *best_component = NULL;
+    opal_accelerator_base_module_t *module = NULL, *best_module = NULL;
+    opal_list_t opened;
+    accelerator_opened_component_t *aoc = NULL;
+
+    /* Traverse the list of available components; call their init
+       functions. */
+
+    OBJ_CONSTRUCT(&opened, opal_list_t);
+    OPAL_LIST_FOREACH(cli, &opal_accelerator_base_framework.framework_components, mca_base_component_list_item_t) {
+        component = (opal_accelerator_base_component_t *) cli->cli_component;
+        if (NULL == component->accelerator_init) {
+            opal_output_verbose( 10, opal_accelerator_base_framework.framework_output,
+                                 "select: no init function; ignoring component %s",
+                                 component->base_version.mca_component_name );
+            continue;
+        }
+
+        /* Init component to get its priority */
+        opal_output_verbose( 10, opal_accelerator_base_framework.framework_output,
+                             "select: initializing %s component %s",
+                             component->base_version.mca_type_name,
+                             component->base_version.mca_component_name );
+        priority = best_priority;
+        module = component->accelerator_init(&priority);
+        if (NULL == module) {
+            opal_output_verbose( 10, opal_accelerator_base_framework.framework_output,
+                                 "select: init returned failure for component %s",
+                                 component->base_version.mca_component_name );
+            continue;
+        }
+        opal_output_verbose( 10, opal_accelerator_base_framework.framework_output,
+                             "select: init returned priority %d", priority );
+        if (priority > best_priority) {
+            best_priority = priority;
+            best_component = component;
+            best_module = module;
+        }
+        aoc = (accelerator_opened_component_t*)malloc(sizeof(accelerator_opened_component_t));
+        if (NULL == aoc) {
+            return OPAL_ERR_OUT_OF_RESOURCE;
+        }
+        OBJ_CONSTRUCT(aoc, opal_list_item_t);
+        aoc->accelerator_component = component;
+        opal_list_append(&opened, (opal_list_item_t*) aoc);
+    }
+
+    /* Finished querying all components. Check for no components found */
+
+    if( NULL == best_component ) {
```

Comment:
why is this an error?  It's also pretty weird that you call show_help, but then continue on through the function.

Reply:
Returned an error code

Reply:
I imagined any component trying to use the accelerator framework will read that it isn't OPAL_SUCCESS (or OPAL_ERR_NOT_FOUND) and will proceed without using the accelerator framework.  I suppose if we move toward having a list of accelerators, it would just be an empty list and any logic should just work with an empty list.

Reply:
I can't think of an example of a non-required framework's select returning an error code if no components are found, but I could be wrong.  Obviously, something like the PML would puke in that situation.  But the onesided and btl componetns would continue with empty lists (until something required a component in that framework, then pukage).  I think the empty list is the best approach; works for both today and any potential multi-component future.